web/src/engine/common/web-utils/src/tests/stringHandling.js

```diff
@@ -0,0 +1,446 @@
+      // assert.equal(str._kmwSubstring(31, -4), str.substring(0, 31));
+      // assert.equal(str._kmwSubstring(-5, 9), str.substring(0, 9));
+      // assert.equal(str._kmwSubstring(-9, -5), '');
+    });
+
+    it('kmwCodePointToCodeUnit', () => {
+      const str = "The quick brown fox jumps over the lazy dog.";
+      assert.equal(str.kmwCodePointToCodeUnit(2), 2);
+      assert.equal(str.kmwCodePointToCodeUnit(9), 9);
+      assert.equal(str.kmwCodePointToCodeUnit(31), 31);
+    });
+
+    it('kmwCodeUnitToCodePoint', () => {
+      const str = "The quick brown fox jumps over the lazy dog.";
+      assert.equal(str.kmwCodeUnitToCodePoint(2), 2);
+      assert.equal(str.kmwCodeUnitToCodePoint(9), 9);
+      assert.equal(str.kmwCodeUnitToCodePoint(31), 31);
+    });
+  });
+
+
+  describe('enabled: SMP-mode, with mixed-plane strings',  () => {
+    before(() => {
+      String.kmwEnableSupplementaryPlane(true);
+    });
+
+    it('_kmwCharAt', () => {
+      // 0x1d5c9: MATHEMATICAL SANS-SERIF SMALL p
+      // 0x1d5be: MATHEMATICAL SANS-SERIF SMALL e
+      const apples = 'a' + String.fromCodePoint(0x1d5c9) + 'p' + 'l' + String.fromCodePoint(0x1d5be) + 's';
+      assert.equal(apples._kmwCharAt(-1), '');
+      assert.equal(apples._kmwCharAt(6), '');
+      assert.equal(apples._kmwCharAt(1000), '');
+      assert.equal(apples._kmwCharAt(0), "a");
+      assert.equal(apples._kmwCharAt(1), String.fromCodePoint(0x1d5c9));
+      assert.equal(apples._kmwCharAt(2), "p");
+      assert.equal(apples._kmwCharAt(5), "s");
+    });
+
+    it('_kmwCharCodeAt', () => {
+      // 0x1d5c9: MATHEMATICAL SANS-SERIF SMALL p
+      // 0x1d5be: MATHEMATICAL SANS-SERIF SMALL e
+      const apples = 'a' + String.fromCodePoint(0x1d5c9) + 'p' + 'l' + String.fromCodePoint(0x1d5be) + 's';
+      assert.isNaN(apples._kmwCharCodeAt(-1));
+      assert.isNaN(apples._kmwCharCodeAt(6));
+      assert.isNaN(apples._kmwCharCodeAt(1000));
+      assert.equal(apples._kmwCharCodeAt(0), "a".charCodeAt(0));
+      assert.equal(apples._kmwCharCodeAt(1), 0x1d5c9);
+      assert.equal(apples._kmwCharCodeAt(2), "p".charCodeAt(0));
+      assert.equal(apples._kmwCharCodeAt(5), "s".charCodeAt(0));
+    });
+
+    it('_kmwIndexOf', () => {
+      // 0x1d5ba: MATHEMATICAL SANS-SERIF SMALL a
+      const a = String.fromCodePoint(0x1d5ba);
+      const banana_nation = "ban" + a + "na n" + a + "tion";
+      assert.equal(banana_nation._kmwIndexOf("XYZ"), -1);
+      assert.equal(banana_nation._kmwIndexOf("na"), 4);
+      assert.equal(banana_nation._kmwIndexOf("na", -1), 4);
+      assert.equal(banana_nation._kmwIndexOf("na", 0), 4);
+      assert.equal(banana_nation._kmwIndexOf("na", 2), 4);
+      assert.equal(banana_nation._kmwIndexOf("na", 3), 4);
+      // ERROR:  returns 4 (!)
+      // assert.equal(banana_nation._kmwIndexOf("na", 5), 7);
+      assert.equal(banana_nation._kmwIndexOf("na", 8), -1);
```

Comment:
`kmwIndexOf` does not properly work for strings with SMP characters when a starting index is specified; this documents one such case.  `kmwLastIndexOf` has the same issue.